Clarify batch and broadcasting behavior in torch.linalg.solve

Neuro1729 opened the pull request:
- Clearly distinguish vector RHS vs. matrix RHS behavior.
- Explain ambiguous cases (e.g., `B.shape = (n, n)`) treated as vector RHS.
- Specify when broadcasting is allowed and when batch dimensions must match.

Fixes #167950

Neuro1729 committed:
Clarify batch and broadcasting behavior in torch.linalg.solve

- Clearly distinguish vector RHS vs. matrix RHS behavior.
- Explain ambiguous cases (e.g., B.shape = (n, n)) treated as vector RHS.
- Specify when broadcasting is allowed and when batch dimensions must match.

lezcano reviewed (commented):
Is this correct? I don't think it's true we disallow broadcasting in the vector case:
```
  const bool vector_case = at::native::linalg_solve_is_vector_rhs(A, B);
  auto B_ = vector_case ? B.unsqueeze(-1) : B;

  // matrix shapes
  at::native::checkInputsSolver(A, B_, /*left=*/left, "linalg.solve");

  // Check that B can be broadcasted to the shape of A
  auto B_broad_shape = std::get<0>(at::native::_linalg_broadcast_batch_dims(B_, A));
```

Neuro1729 committed:
Clarify batch and broadcasting behavior in torch.linalg.solve

hchau630 commented:
Actually maybe this would be an even simpler and clearer way of documenting the batching logic?
```
Letting `*` be zero or more batch dimensions:
- If :attr:`A` has shape `(*, n, n)` and :attr:`B` has shape `(n)` or `(*, n)` (a batch of vectors), 
  this function returns `X` of shape `(*, n)`. The batch dimensions `*` of :attr:`A` and :attr:`B` 
  must **exactly match**.
- Otherwise, if :attr:`A` has shape `(*, n, n)` and :attr:`B` has shape `(n, k)` or `(*, n, k)` (a 
  batch of matrices or “multiple right-hand sides”), this function returns `X` of shape `(*, n, k)`. 
  The batch dimensions `*` of :attr:`A` and :attr:`B` need only be broadcastable.
```
This also automatically clarifies the ambiguous case because it shows that the vector case is checked first.

Neuro1729 committed:
Update linalg.solve docs: clarify batching logic for vectors and matrices

Neuro1729 committed:
Update torch/linalg/__init__.py

Co-authored-by: Mario Lezcano Casado <3291265+lezcano@users.noreply.github.com>

Neuro1729 committed:
Remove redundant paragraph in linalg.solve docs

Removed the duplicated paragraph about broadcasting behavior of B,
as it is already covered by the updated vector and matrix RHS bullet points
requested by the reviewers.

lezcano reviewed (dismissed):
LGTM, thank you! @hchau630 can you confirm you are happy with the result?

Neuro1729 committed:
Fix linting errors in linalg docs

Neuro1729 reviewed (commented):
Hey, I’m done, please take a look

lezcano commented:
There are still linting issues. Please see

>You can reproduce these results locally by using `lintrunner -m origin/main`. (If you don't get the same results, run \'lintrunner init\' to update your local linter)
>See https://github.com/pytorch/pytorch/wiki/lintrunner for setup instructions. To apply suggested patches automatically, use the -a flag. Before pushing another commit,
